### ISSUE TITLE
Native runtime error due to HibernateBasicProxy

### ISSUE THREAD

**Issue by gnieser** [quarkusio/quarkus#16860]
## Describe the bug
Hello,
When using a small hierarchy of JPA entities (abstract parent and concrete child plus discriminator), Hibernate generates a proxy entity that does not seem to be kept by the quarkus-hibernate-orm extension during native compilation.

### Expected behavior
The entity should be loaded from database when in native mode. This is the observed behavior with the uber-jar packaging.

### Actual behavior
When trying to retrieve an entity, the following error occurs:
```
Caused by: java.lang.InstantiationException: Type `org.example.quarkus.AbstractEntity$HibernateBasicProxy$SwF3dMZa` can not be instantiated reflectively as it does not have a no-parameter constructor or the no-parameter constructor has not been added explicitly to the native image.
at java.lang.Class.newInstance(DynamicHub.java:912)
at org.hibernate.bytecode.internal.bytebuddy.BasicProxyFactoryImpl.getProxy(BasicProxyFactoryImpl.java:57)
```

## To Reproduce
Please find attached a reproducer:
[quarkus-hibernateproxy-reproducer.zip](https://github.com/quarkusio/quarkus/files/6389443/quarkus-hibernateproxy-reproducer.zip)

1. Build the native executable
```
mvn clean package -Pnative
```
2. Start the external standalone H2 database
```
java -cp ./h2/h2-1.4.199.jar org.h2.tools.Server -ifNotExists &
```
3. Start the quarkus native executable 
```
./target/quarkus-hibernateproxy-reproducer-0.0.1-SNAPSHOT-runner
```
4. Access the GET /entities endpoint
```
curl localhost:8080/entities
```

### Configuration
This is the application.properties of the above reproducer. Real world usage is Oracle database (btw, many thanks for this new feature).
```properties
quarkus.datasource.db-kind=h2
quarkus.datasource.jdbc.url=jdbc:h2:tcp://localhost/./h2/quarkus
quarkus.datasource.jdbc.driver=org.h2.Driver
quarkus.hibernate-orm.dialect=org.hibernate.dialect.H2Dialect
quarkus.hibernate-orm.statistics=true
quarkus.hibernate-orm.log.sql=true
quarkus.hibernate-orm.database.generation=drop-and-create
quarkus.hibernate-orm.sql-load-script=import.sql
```
This configuration and the -ifNotExists flag on the h2 command line will result in the creation of a local database file
`quarkus-hibernateproxy-reproducer\h2\quarkus.mv.db` which can be deleted between test runs.

## Environment (please complete the following information):
Real world is OpenShift. Reproducer details follows.

### Output of `uname -a` or `ver`
`Linux rhel.domain.local 4.18.0-240.22.1.el8_3.x86_64 #1 SMP Thu Mar 25 14:36:04 EDT 2021 x86_64 x86_64 x86_64 GNU/Linux`

### Output of `java -version`
```
openjdk version "11.0.11" 2021-04-20
OpenJDK Runtime Environment GraalVM CE 21.1.0 (build 11.0.11+8-jvmci-21.1-b05)
OpenJDK 64-Bit Server VM GraalVM CE 21.1.0 (build 11.0.11+8-jvmci-21.1-b05, mixed mode, sharing)
```

### Quarkus version or git rev
1.13.2.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)
```
Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Maven home: /root/apache-maven-3.6.3
Java version: 11.0.11, vendor: GraalVM Community, runtime: /root/graalvm-ce-java11-21.1.0
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "4.18.0-240.22.1.el8_3.x86_64", arch: "amd64", family: "unix"
```

### native-image-agent
Running the uber-jar with the native-image-agent
```
java -agentlib:native-image-agent=config-output-dir=/quarkus/agent -jar target/quarkus-hibernateproxy-reproducer-0.0.1-SNAPSHOT-native-image-source-jar/quarkus-hibernateproxy-reproducer-0.0.1-SNAPSHOT-runner.jar
```
Generates a reflect-config.json file that contains reference to the culprit proxy class
```
$ grep -B1 -A2 BasicProxy /quarkus/agent/reflect-config.json
{
  "name":"org.example.quarkus.AbstractEntity$HibernateBasicProxy$SwF3dMZa",
  "methods":[{"name":"<init>","parameterTypes":[] }]
},
```
### zulip
https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/Native.20runtime.20error.20due.20to.20HibernateBasicProxy

**Comment by Sanne** [comment]
This will be fixed by https://hibernate.atlassian.net/browse/HHH-14575

**Comment by Sanne** [comment]
I've merged HHH-14575 in Hibernate ORM's branches `5.5.x` and `5.4.x`, this will be resolved when we upgrade Quarkus to the next bugfix release - but let's also add tests in Quarkus for this case as I can't easily add native-image tests in Hibernate ORM.
